tt_metal/tt_metal.cpp

```diff
@@ -364,14 +364,27 @@ void CloseDevices(std::map<chip_id_t, Device *> devices) {
             for (uint32_t ts = t.size() - 1; ts > 0; ts--) {
                 if (devices.find(t[ts]) != devices.end()) {
                     devices[t[ts]]->close();
+                    // When a device is closed, its worker thread is joined. Stop tracking this
+                    // worker thread.
+                    tt::DevicePool::instance().unregister_worker_thread_for_device(devices[t[ts]]);
                 }
             }
         }
         //finally close the mmio device
         dev->close();
+        tt::DevicePool::instance().unregister_worker_thread_for_device(dev);
     }
 }
 
+bool InApplicationThread() {
+    bool in_application_thread = true;
+    if (tt::DevicePool::is_instantiated()) {
+        auto worker_thread_ids = tt::DevicePool::instance().get_worker_thread_ids();
```

Comment:
As discussed offline, please review this part

Reply:
As we discussed, we have a race if the user does something like this:

```
Main Thread: CreateDevices();
Spawn an Additional Application Thread
     Application Thread: Start a workload inside the application thread
Main Thread: devices.enable_async();
```

To detect or bypass a race in this scenario, I think the cleanest thing to do is pass an `enable_async` flag to the device/device_pool constructor. This is a fairly invasive change that touches model code. I'll file an issue for this and make the required changes in a separate PR, to unblock current efforts.

Reply:
You are likely correct about the fix.

Additionally, I want to point out that in some scenarios, it can be beneficial to acquire an instance of a resource via a lock guard. This approach ensures that only one thread can access the instance of the pool at any given time, preventing concurrent modifications. In this model, the instance() method should return a guarded pool instance, where the lock is automatically released when the guard goes out of scope. This ensures that while one thread is working with the pool, other threads must wait until the lock is released.

This model is acceptable when there is no need to maintain a persistent reference to the pool and when the methods invoked on the pool are short-lived and interact minimally with shared data structures.

However, this approach would not be ideal if the DevicePool object contains lengthy methods that only access shared structures for a brief period. In such cases, the extended lock duration could lead to unnecessary stalls, affecting overall performance.

Reply:
Sounds good, its a much simpler and more flexibile programming model this way. I can profile host side perf with a mutex in place to verify that the lock doesn't affect performance significantly. If so, this is likely a much better approach. Thanks!

Reply:
Filed a ticket to track this here: https://github.com/tenstorrent/tt-metal/issues/13567